R/match_on_ltcs.R

```diff
@@ -0,0 +1,171 @@
+        # If they have a maternity record, assume the younger age
+        chi_age_max > 3 & recid == "02B" ~ chi_dob_max,
+        # If they have a GLS record and the age is broadly correct, assume older
+        dplyr::between(chi_age_max, 50, 130) & recid == "GLS" ~ chi_dob_min,
+        # If a congenital defect lines up with a dob, assume it is correct
+        chi_dob_max == congen_date ~ chi_dob_max,
+        chi_dob_min == congen_date ~ chi_dob_min,
+        # If being older makes them over 113, assume they are younger
+        chi_age_max > 113 ~ chi_dob_max
+      ),
+      # Create a new variable that explains why the dob and age were changed or not changed
+      changed_dob = dplyr::case_when(
+        # Case when the dob is already valid
+        .data$dob == chi_dob_min ~ 1,
+        .data$dob == chi_dob_max ~ 1,
+        # If one option for dob isn't there, use the other
+        is.na(chi_dob_min) & !is.na(chi_dob_max) ~ 2,
+        !is.na(chi_dob_min) & is.na(chi_dob_max) ~ 2,
+        # If the age would be negative, use the older
+        chi_age_min < 0 ~ 3,
+        # If they have activity before birth date, assume older
+        chi_dob_max > keydate1_dateformat ~ 4,
+        # If they have an LTC date before birth date, assume older
+        chi_dob_max > purrr::reduce(dplyr::select(., arth_date:digestive_date), `min`) ~ 5,
+        # If they have a maternity record, assume the younger age
+        chi_age_max > 3 & recid == "02B" ~ 6,
+        # If they have a GLS record and the age is broadly correct, assume older
+        dplyr::between(chi_age_max, 50, 130) & recid == "GLS" ~ 7,
+        # If a congenital defect lines up with a dob, assume it is correct
+        chi_dob_max == congen_date ~ 8,
+        chi_dob_min == congen_date ~ 8,
+        # If being older makes them over 113, assume they are younger
+        chi_age_max > 113 ~ 9
+      )
+    ) %>%
+    # If we still don't have an age, try and fill it in from a previous record.
+    dplyr::mutate(
+      dob = dplyr::case_when(
+        ((is.na(age) & chi == dplyr::lag(chi)) & (chi_age_max == dplyr::lag(age) | chi_dob_min == dplyr::lag(dob))) ~ chi_dob_min,
+        ((is.na(age) & chi == dplyr::lag(chi)) & (chi_age_min == dplyr::lag(age) | chi_dob_max == dplyr::lag(dob))) ~ chi_dob_max,
+        TRUE ~ dob
+      ),
+      age = dplyr::case_when(
+        ((is.na(age) & chi == dplyr::lag(chi)) & (chi_age_max == dplyr::lag(age) | chi_dob_min == dplyr::lag(dob))) ~ chi_age_max,
+        ((is.na(age) & chi == dplyr::lag(chi)) & (chi_age_min == dplyr::lag(age) | chi_dob_max == dplyr::lag(dob))) ~ chi_age_min,
+        TRUE ~ age
+      ),
+      changed_dob = dplyr::case_when(
+        ((is.na(age) & chi == dplyr::lag(chi)) & (chi_age_max == dplyr::lag(age) | chi_dob_min == dplyr::lag(dob))) ~ 10,
+        ((is.na(age) & chi == dplyr::lag(chi)) & (chi_age_min == dplyr::lag(age) | chi_dob_max == dplyr::lag(dob))) ~ 10,
+        TRUE ~ changed_dob
+      )
+    ) %>%
```

Comment:
I'm sure we can do this better in R. e.g. [tidyr::fill](https://tidyr.tidyverse.org/reference/fill.html)
```suggestion
    # If we still don't have an age, try and fill it in from other records.
  dplyr::group_by(chi) %>%
  tidyr::fill(dob, .direction = "downup") %>%
  dplyr::ungroup()
```